### PR TITLE
Fix loss reporting with deepspeed

### LINKED ISSUE #11919: Trainer reported loss is wrong when using DeepSpeed and gradient_accumulation_steps > 1
## Environment info
- `transformers` version: 4.7.0.dev0
- Platform: Windows-10-10.0.19041-SP0
- Python version: 3.8.0
- PyTorch version (GPU?): 1.8.1 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: no but using DeepSpeed on a single node

### Who can help
@stas00, @sgugger (trainer.py)

### See Also
https://github.com/microsoft/DeepSpeed/issues/1107

## Information

Model I am using (Roberta)

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)
* [ ] pretraining a Language Model (wikipedia and bookcorpus datasets)

## To reproduce

1.  run scripts to pretrain a model with DeepSpeed on a single node with 1 GPU for N steps (gradient_accum_steps=1)
2.  run scripts to pretrain a model with DeepSpeed on a single node with 1 GPU for N steps (gradient_accum_steps=8)
3.  note that vast difference in **loss** reported on console by trainer.py

## Expected behavior

reported loss for any number of gradient_accum_steps, nodes, or GPUs should be the mean of all losses; the same order of magnitude as shown when training with gradient_accum_steps=1, on a single node, with a single GPU.

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

In the deepspeed integration inside `Trainer`, the loss currently reported is the scaled loss, as in scaled by the loss scaling factor used during mixed precision training. This has nothing to do with the actual loss (the scaling factor is the highest possible value that does not make the gradients overflow basically), this PR fixes that.

Fixes #11919

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1750,8 +1750,9 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
             with amp.scale_loss(loss, self.optimizer) as scaled_loss:
                 scaled_loss.backward()
         elif self.deepspeed:
-            # loss gets scaled under gradient_accumulation_steps in deepspeed
-            loss = self.deepspeed.backward(loss)
```
Comment (sgugger) [inline src/transformers/trainer.py:1754]: The value returned here is the scaled loss like the `scaled_loss` above.

**Comment by stas00** [comment]
This is exactly what deepspeed already does. Please see:

https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L1142-L1143

The scaling function is:
```
    def _scale_loss_by_gas(self, prescaled_loss):
        if isinstance(prescaled_loss, torch.Tensor):
            scaled_loss = prescaled_loss / self.gradient_accumulation_steps()
        elif isinstance(prescaled_loss, tuple) or isinstance(prescaled_loss, list):
            scaled_loss = []
            for l in prescaled_loss:
                if isinstance(l, torch.Tensor):
                    scaled_loss.append(l / self.gradient_accumulation_steps())
                else:
                    scaled_loss.append(l)
        else:
            scaled_loss = prescaled_loss
            if self.warn_unscaled_loss:
                logger.warning(
                    f'DeepSpeed unable to scale loss because of type: {type(prescaled_loss)}'
                )
                self.warn_unscaled_loss = False

        return scaled_loss
```

**Comment by stas00** [comment]
I spent some more time running tests, including fp16, and I can't find any problem with the current code. 

As posted in a comment above reviewing the code shows that it only scales by grad acc steps.
